packages/react/src/time-input/TimeInput.tsx

```diff
@@ -0,0 +1,121 @@
+import {
+	forwardRef,
+	type ChangeEvent,
+	type FocusEvent,
+	useEffect,
+	useMemo,
+	useState,
+} from 'react';
+import { Field } from '../field';
+import { TextInputProps, textInputStyles } from '../text-input';
+import {
+	acceptedTimeFormats,
+	formatTime,
+	transformValuePropToInputValue,
+	type TimeFormat,
+} from './utils';
+
+// This const object is required for the tests and also makes the component slightly more performant
+const STABLE_DEFAUT_VALUE: TimeValue = {
+	formatted: '',
+	raw: '',
+};
+
+export const TimeInput = forwardRef<HTMLInputElement, TimeInputProps>(
+	function TimeInput(
+		{
+			block,
+			hideOptionalLabel,
+			hint,
+			id,
+			invalid,
+			label,
+			maxWidth = 'md',
+			message,
+			onBlur: onBlurProp,
+			onChange: onChangeProp,
+			required,
+			timeFormat: timeFormatProp = 'h:mm aaa',
+			value = STABLE_DEFAUT_VALUE,
+			...props
+		},
+		ref
+	) {
+		const [inputValue, setInputValue] = useState(
+			transformValuePropToInputValue(value?.raw, timeFormatProp)
+		);
+		// Update the input when the value and/or timeFormat updates
+		useEffect(() => {
+			setInputValue(transformValuePropToInputValue(value.raw, timeFormatProp));
+		}, [timeFormatProp, value]);
+
+		const onBlur = (e: FocusEvent<HTMLInputElement>) => {
+			const inputValue = e.target?.value;
+			const normalizedTime = formatTime(inputValue, 'HH:mm');
+
+			onChangeProp?.({
+				formatted: formatTime(inputValue, timeFormatProp),
+				raw: normalizedTime,
+			});
+
+			onBlurProp?.(e);
+		};
+
+		const onChange = (e: ChangeEvent<HTMLInputElement>) => {
+			const inputValue = e.target?.value;
+			// Immediately update the input field
+			setInputValue(inputValue);
+		};
+
+		const secondaryLabel = useMemo(() => {
+			const timeFormat = acceptedTimeFormats.includes(timeFormatProp)
+				? timeFormatProp
+				: 'h:mm aaa';
+			return '(e.g. ' + formatTime('21:30', timeFormat) + ')';
+		}, [timeFormatProp]);
+
+		const styles = textInputStyles({ block, maxWidth });
+
+		return (
+			<Field
+				hideOptionalLabel={hideOptionalLabel}
+				hint={hint}
+				id={id}
+				invalid={invalid}
+				label={label}
+				maxWidth={maxWidth}
+				message={message}
+				required={Boolean(required)}
+				secondaryLabel={secondaryLabel}
+			>
+				{(a11yProps) => (
+					<input
+						{...a11yProps}
+						css={styles}
+						onBlur={onBlur}
+						onChange={onChange}
+						ref={ref}
+						type="text"
+						value={inputValue}
+						{...props}
+					/>
+				)}
+			</Field>
+		);
+	}
+);
+
+export type TimeValue =
+	| {
+			/** The time in the format that is displayed to the user **/
+			formatted?: string;
+			/** The time in a consistent 24 hour format **/
+			raw: string;
+	  }
```

Comment:
This should be separated into 2 types and exported as `TimeObject` and `TimeValue`.

```ts
export interface TimeObject {
	/** The time in the format that is displayed to the user **/
	formatted?: string;
	/** The time in a consistent 24 hour format **/
	raw: string;
}
```

```ts
export type TimeValue = TimeObject  | undefined
```

This makes it much easier for people to type their code correctly if they need to type something as TimeValue without `undefined`

Reply:
But then your union is a reference rather than a literal and in order to see its content you need to click into it, rather than just hovering it. You could also wrap the type in `NonNullable` if you really want to remove the `undefined`. I think what's here is better.

Reply:
Here's a playground link. If you hover the types you can see the difference I'm talking to.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAKglgWwgNQIYBsCu0C8BYAKAEgAfKAb0KOoHoAqO2AC2mEWjgDspgWoAzAPYAnBKmA8m4qHADOUACZyw6VCAgKegydEyyIwqAxpVqQ0eOAaA-AC4os4MK4BzANymi9RjD5skMtyoUADGgpyyclacEgBMACxQTIKYhuZiEsaewqgA7vaOzpzuVFBQAL6EZWVkmJwKEPxcGh4EhFxWwvyoIdDwSADyAEYAVhAhEpTE3sys7IE6AiIZktJyisqq6prA2ry6+oZZxOmWNgVOrq1eDLM881xQwWERURAxUAlJKWnL0sdEHL5ByXYqtSptAigSCwdhoLAQACiAA9og1NDhYYNRuMJFBavVGs0FK1CC9HEtBPZ+igMNgoJjyFAgfYAORCQSsiqkgjkiRDVDCansYZjCYMijMvJsgXCLnlHl8qACgBe9gAcuF1Zh0KohugIAAeGnw7AAPglTJZUFZqvlbiAA

Reply:
...I'm a bit torn on this now.

On the one hand, I agree that it is useful being able to hover over `value` and see the format of the object.

On the other hand, it encourages a bad coding practice of using objects directly in union types instead of setting up separate types for each object.

We need to at least have a comment saying "use `NonNullable<TimeValue>` to remove the "undefined" type"

Reply:
>On the other hand, it encourages a bad coding practice of using objects directly in union types instead of setting up separate types for each object.

I don't think it's bad coding practice. It's quite readable as it's a very small object. I would argue that the abstraction of types in this scenario reduces readability.

>We need to at least have a comment saying "use NonNullable<TimeValue> to remove the "undefined" type"

We should be careful about explaining how to write TypeScript in docs. That's not really our place.

I honestly think this is a good balance.

Reply:
I honestly don't think this is a blocking comment. You have a very different approach to TypeScript than is in use in this project, and that's fine, but at some point you have to realise that it's not black and white, it's not always bad coding practice and for the majority of people, it's fine